### PR TITLE
Flow_framework spec refactor: Added Tests, Param, and Error Types

### PR DESCRIPTION
**Pull request by junweid62**
### Description
1. Added missing param for update workflow request body
2. Added more 4xx Errors:
   - TemplateNameRequiredError
   - InvalidTemplateVersionError
   - UnsupportedFieldUpdateError
   - WorkflowParsingError
3. Added test for workflow sample template 

### Issues Resolved
Task two of [#833](https://github.com/opensearch-project/flow-framework/issues/833)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by junweid62** [comment]
Hey @dbwiddis just created a new pr, please take a look when you get a chance :)

File: tests/default/flow_framework/sample_templates_test/deploy_openai_model.yaml [hunk 1]
```diff
@@ -0,0 +1,264 @@
+$schema: ../../../../json_schemas/test_story.schema.yaml
```
Comment (dblock) [inline tests/default/flow_framework/sample_templates_test/deploy_openai_model.yaml:1]: Let's organize these to match the API route like other tests. This is a workflow that uses a template. The descriptions can talk about the OpenAI model.

Maybe `tests/default/flow_framework/workflow/template.yaml`.

File: tests/default/flow_framework/sample_templates_test/deploy_openai_model.yaml [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    response:
+      status: 200
+  - synopsis: Delete workflow.
+    path: /_plugins/_flow_framework/workflow/{workflow_id}
+    method: DELETE
+    parameters:
+      workflow_id: ${create_flow_framework.test_workflow_id}
+    response:
+      status: 200
+  - synopsis: Delete workflow using an invalid ID.
```
Comment (dblock) [inline tests/default/flow_framework/sample_templates_test/deploy_openai_model.yaml:256]: This is already tested in `workflow.yaml`, so probably unneeded here.

**Review by dblock — APPROVED** [review]
Looks good. I have some small asks to consider, please?

**Comment by dblock** [comment]
Appreciate if you could suppress these warnings too.

```
PASSED  flow_framework/deprovision.yaml (tests/default/flow_framework/deprovision.yaml)
WARNING Multiple paths detected, please group similar tests together and move paths not being tested to prologues or epilogues.
  /_plugins/_flow_framework/workflow/{workflow_id}/_status
  /_plugins/_flow_framework/workflow/{workflow_id}/_deprovision

PASSED  flow_framework/provision.yaml (tests/default/flow_framework/provision.yaml)
WARNING Multiple paths detected, please group similar tests together and move paths not being tested to prologues or epilogues.
  /_plugins/_flow_framework/workflow/{workflow_id}/_status
  /_plugins/_flow_framework/workflow/{workflow_id}/_provision
```

I believe `_status` can be moved to a prologue, or may not actually be necessary at all. Generally try to keep these tests to the minimum that exercises the API being tested. This is because we're going to be using them in documentation examples.

**Comment by junweid62** [comment]
@dblock Hi, just pushed a new version addressed all the comment :)

**Comment by junweid62** [comment]
@dblock fixed the link check and pushed a new version. :)

**Review by dblock — APPROVED** [review]

File: tests/default/flow_framework/template.yaml [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    path: /_plugins/_flow_framework/workflow/{workflow_id}
+    method: PUT
+    parameters:
+      workflow_id: invalid
+    request:
+      payload:
+        name: test_update_work_flow
+    response:
+      status: 404
+  - synopsis: Search workflow state.
```
Comment (dblock) [inline tests/default/flow_framework/template.yaml:221]: In general you want to break this one up too into `tests/default/flow_framework/state/search.yaml` with just enough prerequisites. It does feel like repetition, but we will be generating doc pages for each API and we'll want an example that only shows *that* API and not anything else.

**Pull request merged**